examples/experimental/terrain/src/terrain-layer/terrain-layer.js

```diff
@@ -1,124 +1,147 @@
+  updateState({props, oldProps}) {
+    if (props.terrainImage !== oldProps.terrainImage) {
+      const isTiled = props.terrainImage.includes('{x}') && props.terrainImage.includes('{y}');
+      this.setState({isTiled});
     }
-  }
-  // backfill bottom border
-  for (let i = gridSize * (gridSize - 1), x = 0; x < gridSize - 1; x++, i++) {
-    terrain[i] = terrain[i - gridSize];
-  }
-  // backfill right border
-  for (let i = gridSize - 1, y = 0; y < gridSize; y++, i += gridSize) {
-    terrain[i] = terrain[i - 1];
-  }
-  return terrain;
-}
 
-function getMeshAttributes(vertices, terrain, tileSize, bounds) {
-  const gridSize = tileSize + 1;
-  const numOfVerticies = vertices.length / 2;
-  // vec3. x, y in pixels, z in meters
-  const positions = new Float32Array(numOfVerticies * 3);
-  // vec2. 1 to 1 relationship with position. represents the uv on the texture image. 0,0 to 1,1.
-  const texCoords = new Float32Array(numOfVerticies * 2);
-
-  const [minX, minY, maxX, maxY] = bounds;
-  const xScale = (maxX - minX) / tileSize;
-  const yScale = (maxY - minY) / tileSize;
-
-  for (let i = 0; i < numOfVerticies; i++) {
-    const x = vertices[i * 2];
-    const y = vertices[i * 2 + 1];
-    const pixelIdx = y * gridSize + x;
-
-    positions[3 * i + 0] = x * xScale + minX;
-    positions[3 * i + 1] = -y * yScale + maxY;
-    positions[3 * i + 2] = terrain[pixelIdx];
-
-    texCoords[2 * i + 0] = x / tileSize;
-    texCoords[2 * i + 1] = y / tileSize;
+    // Reloading for single terrain mesh
+    const shouldReload =
+      props.meshMaxError !== oldProps.meshMaxError ||
+      props.elevationDecoder !== oldProps.elevationDecoder ||
+      props.bounds !== oldProps.bounds;
+
+    if (!this.state.isTiled && shouldReload) {
+      const options = {
+        terrain: {
+          meshMaxError: props.meshMaxError,
+          elevationDecoder: props.elevationDecoder,
+          bounds: props.bounds
+        }
+      };
+      this.setState({terrain: load(props.terrainImage, TerrainLoader, options)});
+    }
   }
 
-  return {
-    positions: {value: positions, size: 3},
-    texCoords: {value: texCoords, size: 2}
-    // normals: [], - optional, but creates the high poly look with lighting
-  };
-}
-
-function getMartiniTileMesh(terrainImage, getElevation, meshMaxError, bounds) {
-  if (terrainImage === null) {
-    return null;
+  getTerrainData({bbox, x, y, z}) {
+    const {terrainImage, elevationDecoder, meshMaxError} = this.props;
+    const url = terrainImage
+      .replace('{x}', x)
+      .replace('{y}', y)
+      .replace('{z}', z);
+
+    const viewport = new WebMercatorViewport({
+      longitude: (bbox.west + bbox.east) / 2,
+      latitude: (bbox.north + bbox.south) / 2,
+      zoom: z
+    });
+    const bottomLeft = viewport.projectFlat([bbox.west, bbox.south]);
+    const topRight = viewport.projectFlat([bbox.east, bbox.north]);
+    const bounds = [bottomLeft[0], bottomLeft[1], topRight[0], topRight[1]];
+
+    const options = {
```

Comment:
Extract the construction of options and calling `load` out to a utility function, so that you can reuse in `updateState`.